sklearn/naive_bayes.py

```diff
@@ -958,3 +985,245 @@ def _joint_log_likelihood(self, X):
+
+    def _count(self, X, Y):
+        self.class_count_ += Y.sum(axis=0)
+        for i in range(self.n_features_):
+            X_feature = X[:, i]
+            cats = np.unique(X_feature)
+            self._update_cat_mapping(self.feature_cat_index_mapping_[i], cats)
+            # update category_count_dimensions in case partial_fit is used, to
+            # to account for unseen categories
+            self.category_count_[i] = self._update_cat_count_dims(
+                self.category_count_[i], self.feature_cat_index_mapping_[i])
+            self._update_cat_count(X_feature, Y,
+                                   self.category_count_[i],
+                                   self.feature_cat_index_mapping_[i])
+
+    def _update_cat_mapping(self, cat_mapping, cats):
+        for cat in cats:
+            cat_mapping.setdefault(cat, len(cat_mapping))
+
+    def _update_cat_count_dims(self, cat_count, cat_mapping):
+        diff = len(cat_mapping) - cat_count.shape[1]
+        if diff > 0:
+            return np.pad(cat_count, [(0, 0), (0, diff)], 'constant')
+        return cat_count
+
+    def _update_cat_count(self, X_feature, Y, cat_count, cat_mapping):
+        for j in range(self.class_count_.shape[0]):
+            X_feature_class = X_feature[Y[:, j] > 0]
+            Y_class = Y[Y[:, j] > 0, j]
+            class_cats, n_feature_class = self._unique_sums(X_feature_class,
+                                                            Y_class)
+            indices = [cat_mapping[cat] for cat in class_cats]
+            cat_count[j, indices] += n_feature_class
+
+    def _unique_sums(self, X_feature_class, Y_class):
+        # Sum for each category in x the corresponding values in y
+        class_cats, inv = np.unique(X_feature_class, return_inverse=True)
+        n_feature_class = np.zeros(len(class_cats), dtype=Y_class.dtype)
+        np.add.at(n_feature_class, inv, Y_class)
+        return class_cats, n_feature_class
+
+    def _update_feature_log_prob(self, alpha):
+        feature_log_prob = {}
+        for i in range(self.n_features_):
+            smoothed_cat_count = self.category_count_[i] + alpha
+            smoothed_class_count = smoothed_cat_count.sum(axis=1)
+            feature_log_prob[i] = (np.log(smoothed_cat_count) -
+                                   np.log(smoothed_class_count.reshape(-1, 1)))
+        self.feature_log_prob_ = feature_log_prob
+
+    def _joint_log_likelihood(self, X):
+        if not X.shape[1] == self.n_features_:
+            raise ValueError("Expected input with %d features, got %d instead"
+                             .format(self.n_features_, X.shape[1]))
+        jll = np.zeros((X.shape[0], self.class_count_.shape[0]))
+        for i in range(self.n_features_):
+            X_feature = X[:, i]
+            indices = []
+            mapping = self.feature_cat_index_mapping_[i]
+            for sample, cat in enumerate(X_feature):
```

Comment:
hm.. looking at the code we could definitely remove `enumerate` since we don't use `sample` anymore. 

What we could do to potentially speed it up a bit and make it more readable is adding a method:
```python
    def _get_cat_indices(self, cat, feature_i):
        try:
            return  self.feature_cat_index_mapping_[feature_i][cat]
        except KeyError:
            if self.handle_unknown == 'warn':
                warnings.warn(
                    "Category {} not expected for feature {} "
                    "of features 0 - {}."
                        .format(cat, feature_i, self.n_features_ - 1)
                )
            elif self.handle_unknown == 'error':
                raise KeyError(
                    "Category {} not expected for feature {} "
                    "of features 0 - {}."
                        .format(cat, feature_i, self.n_features_ - 1)
                )
```
then we would have:
```python
    def _joint_log_likelihood(self, X):
        if not X.shape[1] == self.n_features_:
            raise ValueError("Expected input with %d features, got %d instead"
                             .format(self.n_features_, X.shape[1]))
        jll = np.zeros((X.shape[0], self.class_count_.shape[0]))
        for i in range(self.n_features_):
            indices = [cat for cat in X[:, i] if cat is not None]
            # indices length is 0, if all categories have not been seen in the
            # training set
            if len(indices) > 0:
                jll += self.feature_log_prob_[i][:, indices].T
        total_ll = jll + self.class_log_prior_
        return total_ll
```

According to this [Stackoverflow question](https://stackoverflow.com/questions/35215161/most-efficient-way-to-map-function-over-numpy-array) this is the fastest we can get for applying a general function to each element of an array.

Reply:
I implemented it as @FlorianWilhelm suggested